### PR TITLE
fix(directory): contract git home dirs with symbol (#4024)

### LINKED ISSUE #4024: directory: `$before_root_path` is incorrect
#### Current Behavior
I have my home directory under *Git*. For the `directory` segment, `$before_root_path` is incorrect.

When I have the current config:

```toml
[directory]
truncate_to_repo = true
repo_root_format = '[$before_root_path]($style)'
style = 'fg:white bg:blue'
repo_root_style = 'bold fg:yellow bg:blue'
```

If I got to `~/.config`, then the segment shows:

```
~/.config
```

I expect `$before_root_path` to be nothing here.

My actual config looks like:

```toml
repo_root_format = '[ $before_root_path]($style)[$repo_root]($repo_root_style)[$path ]($style)[$read_only]($read_only_style)'
```

This shows the segment:

```
 ~/.configsteven/.config
```

This gives some path duplication. I hoped this would be fixed by *Starship* eventually, but it's been a bug for many releases now.

#### Environment
- Starship version: 1.7.0
- zsh version: zsh 5.9 (x86_64-pc-linux-gnu)
- Operating system: Arch Linux Unknown
- Terminal emulator: WezTerm 20220408-101518-b908e2dd
- Git Commit Hash: fda3ed8c
- Branch/Tag: v1.7.0
v1.7.1
- Rust Version: rustc 1.61.0 (Arch Linux rust 1:1.61.0-1)
- Rust channel:  release
- Build Time: 2022-05-25 07:01:30 +00:00

### PR DESCRIPTION
**Pull request by mkw**
#### Description
Apply home dir substitution to both the display directory and repository root directory before doing home directory symbol substitution.

#### Motivation and Context
Home dirs stored in git repos resulted in inconsistent path contraciton when the homedir symbol substitution was done. This resulted in output similar to ` ~/.configsteven/.config` instead of just `~/.config` when `$before_root_path` was used.
Closes #4024

#### How Has This Been Tested?
- [x] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

**Comment by stevenxxiu** [comment]
Thanks for this.

One issue I noticed while using this in *Termux*. I have *Git* in my normal home directory, at `/data/data/com.termux/files/home/`.

When I run `tsu` for the root user, and go to the root's home directory, at `/data/data/com.termux/files/home/.suroot/`, the `directory` module crashes, thus displaying an empty directory segment.

I think a reasonable comprimise is to just ignore it when the *Git* directory is an ancestor of the home directory. Here's my patch:

```patch
From 72fb91b3459edbb1790d20cb93f233a6390a536f Mon Sep 17 00:00:00 2001
From: Steven Xu <stevenxxiu@gmail.com>
Date: Sat, 20 May 2023 17:56:07 +1000
Subject: [PATCH] fix(directory): ignore *Git* repos that are parents of the
 home directory

---
 src/modules/directory.rs | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/modules/directory.rs b/src/modules/directory.rs
index c21940c8..4efd50ef 100644
--- a/src/modules/directory.rs
+++ b/src/modules/directory.rs
@@ -105,7 +105,10 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
     };
 
     let path_vec = match &repo.and_then(|r| r.workdir.as_ref()) {
-        Some(repo_root) if config.repo_root_style.is_some() => {
+        Some(repo_root)
+            if config.repo_root_style.is_some()
+                && (**repo_root == home_dir || !home_dir.starts_with(repo_root)) =>
+        {
             // Handle the home directory by doing symbol substitution in both
             // the display directory and the repo root.
             let full_path = contract_path(display_dir, &home_dir, &home_symbol);
-- 
2.40.1

```

**Comment by mkw** [comment]
## Update: Rebased and Enhanced

I've updated this PR with:

1. **Rebased on latest `origin/master`** - Now based on commit `bfd7adea`
2. **Added comprehensive test case** - New test `git_repo_in_home_directory_repo_root_style` verifies the fix works correctly when:
   - Home directory is a git repository
   - `repo_root_style` is configured
   - Working in subdirectories of the home repo

**Fixes #4024**
**Fixes #6513** (duplicate issue reported in 2024)

All directory module tests pass (38 passed, 0 failed, 19 ignored).

The fix ensures both `display_dir` and `repo_root` are contracted with the home symbol before comparison, preventing path duplication like `~/foomyuser/foo` → `~/foo`.

**Comment by mkw** [comment]
## Addressed @stevenxxiu's Termux Concern

Added a safeguard to handle the edge case where a git repository is a strict ancestor (parent) of the home directory. This prevents the crash you encountered in Termux where:

- Git repo: `/data/data/com.termux/files/home/`
- Home dir: `/data/data/com.termux/files/home/.suroot/`

The guard condition now checks:
```rust
(\!home_dir.starts_with(repo_root) || **repo_root == home_dir)
```

This **accepts** repos when:
- The repo does NOT contain home as a descendant (normal case), OR  
- The repo root equals the home directory exactly (the case this PR fixes)

This **rejects** repos when:
- Home is a descendant of the repo but they're not equal (the Termux edge case)

All 38 directory module tests still pass with this safeguard in place.

**Comment by niklaswimmer** [comment]
Can confirm that this indeed solves the issue for me. Picture showing before and after:

image

**Comment by mkw** [comment]
Is there anything that I can do to get this merged? I have been using the core of the fix that I wrote by hand for a few years now, rebasing after each release with no issues. I added tests, and addressed the edge case that @stevenxxiu identified with Claude Code.

I just rebased my fork and re-tested, as well. I confess that I'm not a rust expert, so maybe I need to fix something? Please let me know. I love Starship, but it's really annoying to have to rebuild on every release.

**Comment by that70schris** [comment]
is there any help available for this?

I've started using your fork @mkw, working like a charm on both macOS and Raspberry Pi
